vmm/src/memory_manager.rs

```diff
@@ -485,37 +495,55 @@ impl MemoryManager {
-                        * virtio_devices::VIRTIO_MEM_DEFAULT_BLOCK_SIZE,
-                );
+        if !use_zones {
+            if memory_zones.len() != 1 {
+                return Err(Error::InvalidNumberOfMemoryZones);
+            }
 
-                if !use_zones {
-                    virtiomem_region = Some(MemoryManager::create_ram_region(
-                        &None,
-                        0,
-                        start_addr,
-                        size as usize,
-                        false,
-                        config.shared,
-                        config.hugepages,
-                        None,
-                        &None,
-                    )?);
-                }
+            if let Some(memory_zone) = memory_zones.get_mut(DEFAULT_MEMORY_ZONE) {
+                if let Some(size) = config.hotplug_size {
+                    if size == 0 {
+                        error!("'hotplug_size' can't be 0");
+                        return Err(Error::InvalidHotplugSize);
+                    }
 
-                virtiomem_resize = Some(virtio_devices::Resize::new().map_err(Error::EventFdFail)?);
+                    if config.hotplug_method == HotplugMethod::VirtioMem {
+                        // Alignment must be "natural" i.e. same as size of block
+                        let start_addr = GuestAddress(
+                            (start_of_device_area.0
+                                + virtio_devices::VIRTIO_MEM_DEFAULT_BLOCK_SIZE
+                                - 1)
+                                / virtio_devices::VIRTIO_MEM_DEFAULT_BLOCK_SIZE
+                                * virtio_devices::VIRTIO_MEM_DEFAULT_BLOCK_SIZE,
+                        );
 
-                start_of_device_area = start_addr.unchecked_add(size);
+                        let region = MemoryManager::create_ram_region(
+                            &None,
+                            0,
+                            start_addr,
+                            size as usize,
+                            false,
+                            config.shared,
+                            config.hugepages,
+                            None,
+                            &None,
+                        )?;
+
+                        virtiomem_regions.push(region.clone());
+
+                        memory_zone.virtiomem_region = Some(region);
+                        memory_zone.virtiomem_resize =
+                            Some(virtio_devices::Resize::new().map_err(Error::EventFdFail)?);
+
+                        start_of_device_area = start_addr.unchecked_add(size);
+                    } else {
+                        start_of_device_area = start_of_device_area.unchecked_add(size);
+                    }
+                }
             } else {
-                start_of_device_area = start_of_device_area.unchecked_add(size);
+                return Err(Error::MissingDefaultMemoryZone);
```

Comment:
When can this happen?

Reply:
This is later removed. This was introduced as part of maintaining the behavior for `--memory`, but at some point when introducing support for all zones, this simply goes away.